photosynthesis_metrics/tv.py

```diff
@@ -0,0 +1,153 @@
+r"""Implemetation of Total Variation metric, based on article
+ remi.flamary.com/demos/proxtv.html and www.wikiwand.com/en/Total_variation_denoising
+"""
+
+from typing import Union, Tuple, Optional
+import functools
+import torch
+from torch.nn.modules.loss import _Loss
+
+from .utils import _adjust_dimensions, _validate_input
+
+
+def _compute_tv(x: torch.Tensor, size_average: bool = True, reduction_type: str = 'l2') -> torch.Tensor:
+    r"""Compute Total Variation metric
+    See www.wikiwand.com/en/Total_variation_denoising for additional info about reduction types
+
+    Args:
+        x: Tensor of shape :math:`(N, C, H, W)` holding an input image.
+        size_average: If size_average=True, total variation of all images will be averaged as a scalar.
+        reduction_type: {'l1', 'l2', 'l2_squared'}, defines which type of norm to implement, isotropic  or anisotropic.
+
+    Returns:
+        tv : Total variation of a given tensor
+    """
+
+    assert isinstance(x, torch.Tensor), f'Image stack must be of type torch.Tensor, got {type(x)} instead.'
+
+    if reduction_type == 'l1':
+        w_variance = torch.sum(torch.abs(x[:,:,:,1:] - x[:,:,:,:-1]), dim=[1, 2, 3])
+        h_variance = torch.sum(torch.abs(x[:,:,1:,:] - x[:,:,:-1,:]), dim=[1, 2, 3])
+        tv_val = (h_variance + w_variance)
+    elif reduction_type == 'l2':
+        w_variance = torch.sum(torch.pow(x[:,:,:,1:] - x[:,:,:,:-1], 2), dim=[1, 2, 3])
+        h_variance = torch.sum(torch.pow(x[:,:,1:,:] - x[:,:,:-1,:], 2), dim=[1, 2, 3]) 
+        tv_val = torch.sqrt(h_variance + w_variance)
+    elif reduction_type == 'l2_squared':
+        w_variance = torch.sum(torch.pow(x[:,:,:,1:] - x[:,:,:,:-1], 2), dim=[1, 2, 3])
+        h_variance = torch.sum(torch.pow(x[:,:,1:,:] - x[:,:,:-1,:], 2), dim=[1, 2, 3])  
+        tv_val = (h_variance + w_variance)
+    else:
+        raise ValueError("Incorrect reduction type, should be one of {'l1', 'l2', 'l2_squared'}")
+
+    if size_average:
+        return tv_val.mean()
+    else:
+        return tv_val
+
+def total_variation(x: torch.Tensor, size_average: bool = True, reduction_type: str = 'l2') -> torch.Tensor:
+    r"""Interface of Total Variation.
+
+    Args:
+        x: Batch of images. Required to be 4D, channels first (N,C,H,W).
+        size_average: If size_average=True, total_variation of all images will be averaged as a scalar.
+        reduction_type: one of {'l1', 'l2', 'l2_squared'}
+
+    Returns:
+        Value of Structural Similarity (SSIM) index.
+
+    References:
+        .. [1] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P.
+           (2004). Image quality assessment: From error visibility to
+           structural similarity. IEEE Transactions on Image Processing,
+           13, 600-612.
+           https://ece.uwaterloo.ca/~z70wang/publications/ssim.pdf,
+           :DOI:`10.1109/TIP.2003.819861`
+    """
+    _validate_input(x=x, y=x, kernel_size=3)
```

Comment:
I like your idea to reuse as much code as possible. Please consider removing `kernel_size=3` part since `kernal_size=None` by default and this case is handled within the function.